### PR TITLE
[ClickHouse] Added NoRec Oracle for ClickHouse and other improvements

### PR DESCRIPTION
**Pull request by qoega**

### CODE REVIEW

**Review by mrigger — COMMENTED** [review]
Thanks! Could you have a look at my two comments? As a general comment, when fixing CI test failures like style checks introduced within the PR, could you please amend the changes to the respective commit, or squash all fixes into one, to have a cleaner commit history?

File: src/sqlancer/MainOptions.java [hunk 1]
```diff
@@ -47,6 +47,9 @@ public class MainOptions {
     @Parameter(names = "--log-execution-time", description = "Logs the execution time of each statement (requires --log-each-select to be enabled)", arity = 1)
     private boolean logExecutionTime = true; // NOPMD
 
+    @Parameter(names = "--print-failed", description = "Logs failed insert, create and other statements without results", arity = 1)
+    private boolean loggerPrintFailed = false; // NOPMD
```
[the remark this answers is not in the document]
Comment (qoega) [reply src/sqlancer/MainOptions.java]: There is no indication that shows that one line was noop, another failed etc. I generate some CREATE statements with `IF NOT EXISTS`, other failed and it showed test case that just failed in the middle. It was misleading as I thought I have broken generator that tries to create several tables with the same name.

As an alternative when we reproduce we can just continue after failures. For my case it was easier to skip.

File: src/sqlancer/clickhouse/ClickHouseErrors.java [hunk 2]
```diff
@@ -8,48 +8,52 @@ private ClickHouseErrors() {
...
         errors.add("Sampling expression must be present in the primary key");
         errors.add("Sorting key cannot contain constants");
         errors.add("There is no supertype for types");
+        errors.add("argument of function");
+        errors.add("because it is constant but values of constants are different in source and result"); // https://github.com/ClickHouse/ClickHouse/issues/22119
+        errors.add("but its arguments considered equal according to constraints");
+        errors.add("does not return a value of type UInt8");
+        errors.add("does not return a value of type UInt8");
+        errors.add("doesn't exist"); // TODO: consecutive test runs can lead to dropped database
+        errors.add("in block. There are only columns:"); // https://github.com/ClickHouse/ClickHouse/issues/42399
```
[the remark this answers is not in the document]
Comment (qoega) [reply src/sqlancer/clickhouse/ClickHouseErrors.java:46]: It makes sense to add such flags when they disable some features. Currently just deletion of this line with related issue link will do all the work.

Idea is that generator can return incorrect queries and check how ClickHouse work with them. It is much better in terms of coverage than to pass only correctly built SQL queries. Most of the errors are just simple checks like you can't use String in WHERE condition etc.

[the remark this answers is not in the document]
Comment (mrigger) [reply src/sqlancer/clickhouse/ClickHouseErrors.java:46]: Okay, this is fine, then. Thanks for the explanation.

**Comment by qoega** [comment]
> could you please amend the changes to the respective commit, or squash all fixes into one, to have a cleaner commit history?

I thought "Squash and commit" will do all the work anyway. I split changes into commit just to make it a bit easier to review. They don't have working state in intermediary commits anyway.

File: src/sqlancer/MainOptions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrigger) [reply src/sqlancer/MainOptions.java]: The problem I see is that we don't know whether a statement has a side effect. The return code by the database systems doesn't tell us much in general, so I'm a bit reluctant about this change. Understanding your use case correctly, the automated reducer will also help with your use case. What do you think?

**Comment by mrigger** [comment]
Sounds good, I'll "Squash and merge". Could you still have a look at my follow-up comment on the flag?

**Comment by qoega** [comment]
Yes. It is `true` by default. So it will print the failed statements by default.
https://github.com/sqlancer/sqlancer/pull/598/files#diff-3a91051bb589dac782a7a444130f53997253444fa666ff58edbb3914c18cad31R51

I guess I finally figured out how to all the necessary style checks locally.

**Review by mrigger — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**